Fixes stop and refactors sysfs loading

fiuczy opened the pull request:
The original reason to start this series was to reorganize the loading of the data from the sysfs and separate the loading, error handling and data handling. At the end the data is retrieved into a new structure, the places which require data loading from sysfs decide how to handle errors and what is done with the retrieved data.

Besides that a few other things crossed my path and got changed:
 - adds tests for corrupted mdevs
 - adds checks for testdata dead files
 - contains a few rustification patches
 - tries to handle a race when loading sysfs data and a parallel delete of the mdev

fiuczy committed:
mdev: rename path methods following the same schema

Rename persist_path() to persistent_path() and path() to active_path()
to match the naming schema.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
callouts: report an error encountered in invoke_modify_live

When trying to live modify an mdev the error found when looking up the
mdev in the sysfs should be reported to the user.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
environments: prevent callout error

Loading from sysfs always succeeds and data for mdev_tape might not have
been set. Without checking if the mdev is active the callout fails to
initialize with the error
Error: Device ed4654e5-12f2-4936-b079-4bba74d2c783 must have a defined mdev_type

This problem is caused by raciness between read_dir(), delayed
load_from_sysfs() calls and someone deleting mdevs. The likelihood to
catch this increases with the number of active mdevs and the amount of
time mdev_type matching callout scripts requires to complete
get:attributes callouts.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
main: allow to override a sysfs lookup error on stop

Allow users to force a stop on an mdev which is in a disarranged state.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
tests: adding an unused_file to TestEnvironment

This allows to find unused test files in testdata more easily.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
mdev: separate data retrieval from checking

Introducing a new class for the sysfs data retrieval allows the callers
to make the decision what is going to happen with the retrieved data,
how to handle encountered errors and what checking needs happen before
storing the data into a mdev object.

Adding the methods set_sysfs_data for storing and is_sysfs_data_matching
for checking to the mdev class for easy usability.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
mdev: report error on create with warning

We already know that something is suspicious with the uuid used to
create a new mdev. Instead of ignoring the error report it with a
warning to the user.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
environment: ignore sysfs error in get_defined_devices

When defined mdevs are retrieved a sysfs look failure should not be
disruptive. Provide a warning instead.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
tests: add tests with broken active mdevs

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
mdev: remove panic in load_from_sysfs and tighten data retrieval

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
main: allow define to proceed if sysfs lookup fails

Allow to force a define an mdev if the sysfs data retrieval for the mdev
failed. The user will receive a warning.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
main: fix stop when device not active

Trying to stop a not existing mdev causes the following error:
Error: Device ed4654e5-12f2-4936-b079-4bba74d2c783 must have a defined mdev_type

This is a message occurs when the callout object is initiated and the
mdev has no mdev_type set.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
mdev: add check in load_from_sysfs for parallel device removal

Since we already checked in the sysfs that mdev exists the failures to
retrieve parent and type can either be caused by
 a) the mdev was delete in the meantime or
 b) an unknown reason
From case a) we can actually recover into a nearly data consistent way
by rechecking if the mdev was deleted. For case b) we must report the
error as before.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
mdev: canonical_basename rustification

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy committed:
tests: fix clippy warnings

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy commented:
I added another patch which fixes clippy warnings in the tests code and resolved the resulting merge conflicts as well as added changes in the new code to remove clippy warnings.

jonner commented on src/tests/mod.rs:
```diff
@@ -247,27 +247,21 @@ impl TestEnvironment {
-        .and_then(|_| {
-            println!("Regenerated expected data file {:?}", filename);
-            Ok(())
-        })
```
I think maybe we just remove the whole function chaining here instead:

```rust
    fs::write(filename, data.as_bytes())?;
    println!("Regenerated expected data file {:?}", filename);
    Ok(())
```

fiuczy replied:
Ok, changing it.

jonner reviewed (commented):
First patch looks fine with one minor suggestion. I don't have these clippy warnings so you must have been using a different version of clippy.

fiuczy commented on src/tests/modify.rs:
```diff
@@ -673,7 +667,9 @@ fn test_modify() {
-            format!("'live' option must be used with 'jsonfile' option").as_str(),
+            "'live' option must be used with 'jsonfile' option"
+                .to_string()
+                .as_str(),
```
I will remove lines 671 and 672.

fiuczy commented:
> First patch looks fine with one minor suggestion. I don't have these clippy warnings so you must have been using a different version of clippy.

You should be able to get to the messages with `cargo clippy --all-targets --all-features -- -D warnings`.

jonner reviewed (commented):
For patch 2 ("add tests with broken active mdevs"), I'm a little bit confused. Did you actually run into broken mdevs in sysfs during normal operation?  And if so, was mdevctl doing the wrong thing in this case? This seems to only add tests cases and doesn't seem to modify behavior...

jonner commented on src/mdev.rs:
```diff
@@ -102,14 +102,25 @@ impl MDev {
+                .parent()
+                .ok_or_else(|| anyhow!("Path to parent of mdev {:?} does not exist", self.uuid))?;
+            canonical_basename(sysfsparent)
+        };
```
These closures seem unnecessary and indirect. We could simply replace the `unwrap()` call below directly with an `ok_or_else()?` and achieve the same thing with a single line change.

fiuczy replied:
This happens in later patches as this patch is a preparation for further changes.

jonner reviewed (commented):
In the patch "add check in load_from_sysfs for parallel device removal":

We might as well just remove the initial `path.exists()` check if we're just going to check it again anyways. And in fact, I'd prefer to not call `Path::exists()` at all. `Path::canonicalize()` already traverses all symlinks, so I think we can simply match against the error returned by `Path::canonicalize()` and check if it is a `std::io::ErrorKind::NotFound`.

jonner commented:
I see now that the behavior tested in patch 2 was modified by later commits...

jonner commented:
To me, this `MDevSysfsData` object feels a bit awkward. It feels like it would be nicer if the `MDev` object had separate 'active config' and a 'defined config' members rather than having this standalone object that can be queried and then used to set values on an `MDev` object. Did you try something like that?

jonner commented on src/mdev.rs:
```diff
@@ -16,6 +16,94 @@ pub enum FormatType {
+}
+
+impl MDevSysfsData {
+    pub fn new(env: &Rc<dyn Environment>, uuid: &Uuid) -> Result<MDevSysfsData> {
```
to me, `new()` indicates a simple allocation, not a bunch of I/O that can fail. So something like `load()` would be a much better name.

fiuczy replied:
Ok, I will rename it.

jonner commented on src/mdev.rs:
```diff
@@ -16,6 +16,94 @@ pub enum FormatType {
 
+pub struct MDevSysfsData {
+    pub uuid: Uuid,
+    pub active: bool,
+    pub parent: Option<String>,
+    pub mdev_type: Option<String>,
```
I don't think it makes sense to have an `active` field here or `Option`s for the other fields. 

```suggestion
    pub parent: String,
    pub mdev_type: String,
```

If the device is not active, the `load()` function it should simply not return a `MDevSysfsData` object. Having all of these fields be separately nullable implies that it might be valid to have an object where `active` is true, and `parent` is `None`, etc.

fiuczy replied:
See my comment on `new()`. 
`active` can be `false` and `parent `as well as m`dev_type` are than `None`.
Maybe the struct fields should not be public and only accessible via getters.

jonner commented on src/mdev.rs:
```diff
@@ -16,6 +16,94 @@ pub enum FormatType {
+            parent = match Self::load_parent_from_sysfs(&active_path) {
+                Ok(parentname) => Some(parentname),
+                Err(e) => {
+                    if Self::is_active(&active_path) {
```
As I said in an eralier comment, checking for `is_active()` before, then calling `load_parent_from_sysfs()`, and then calling `is_active()` again is totally unnecessary. the `load_parent_from_sysfs()` function should already return an error if the file doesn't exist, so you should just be able to call that directly and then check the error type.

fiuczy replied:
It seems like it is unnecessary but it is not.
The first `is_active()` is done to check right before getting the parent data from the sysfs.
In case the `load_parent_from_sysfs()` fails this could be for multiple reasons BUT if in parallel the mdev is deleted between the first `is_active()` and the `load_parent_from_sysfs()` this should not cause an error but instead simply be handled as "device is inactive". Same is true for the following `load_mdev_type_from_sysfs()` call.
We have seen this happen in our CI for a few month now.

fiuczy replied:
In patch [mdev: MDevSysfsData.load() without path.exists() check](https://github.com/mdevctl/mdevctl/pull/119/commits/96812712ac6fcc74d944614fe6714af17efa6c63) `is_active()` gets removed.

jonner commented on src/main.rs:
```diff
@@ -413,7 +427,23 @@ fn start_command(
     debug!("Stopping '{}'", uuid);
     let mut dev = MDev::new(env, uuid);
-    dev.load_from_sysfs()?;
+    match MDevSysfsData::new_with_mdev(&dev) {
+        Ok(sysfs_data) => {
+            if !sysfs_data.active {
+                return Err(anyhow!("Device {} is not an active mdev", uuid));
+            }
+            dev.set_sysfs_data(sysfs_data);
+        }
+        Err(e) => {
+            if !force {
+                return Err(e);
+            }
+            warn!(
+                "For device {} a sysfs update caused the error: {:?}",
+                dev.uuid, e
+            );
+        }
+    };
```
This is about the 3rd time I've seen this same basic pattern repeated. It feels a bit awkward and verbose.

fiuczy replied:
Sure, error handling has a basic pattern but the details differ in every case a bit.

jonner commented on src/mdev.rs:
```diff
@@ -16,6 +16,94 @@ pub enum FormatType {
+        })
+    }
+
+    pub fn new_with_mdev(mdev: &MDev) -> Result<MDevSysfsData> {
```
same comment about naming: `new` vs. `load`

fiuczy replied:
OK

fiuczy commented:
> For patch 2 ("add tests with broken active mdevs"), I'm a little bit confused. Did you actually run into broken mdevs in sysfs during normal operation? And if so, was mdevctl doing the wrong thing in this case? This seems to only add tests cases and doesn't seem to modify behavior...

Yes, patch 2 adds only tests which cover the scenario that the mdev is somehow in an incorrect state.
There are no changes in mdevctl behavior itself. These changes are introduced in later patches.

We have seen problems with the mdevctl code which reads data from sysfs when testing with many mdevs.
The scenarios causing errors resulted from mdevctl reading sysfs for mdev x and in parallel the mdev x is deleted.
We have not found mdevctl creating invalid mdevs.

fiuczy commented on src/mdev.rs:
```diff
@@ -118,8 +118,28 @@ impl MDev {
             debug!("device did not exist in sysfs: {:?}", self);
```
> In the patch "add check in load_from_sysfs for parallel device removal":
>
> We might as well just remove the initial path.exists() check if we're just going to check it again anyways. And in fact, I'd prefer to not call Path::exists() at all. Path::canonicalize() already traverses all symlinks, so I think we can simply match against the error returned by Path::canonicalize() and check if it is a std::io::ErrorKind::NotFound.

I can give it a try as I am not sure if the error does match only the link to the devices directory no longer exists. I am also not sure if this improves readability but let me try it out.

fiuczy replied:
I added patch [mdev: MDevSysfsData.load() without path.exists() check](https://github.com/mdevctl/mdevctl/pull/119/commits/96812712ac6fcc74d944614fe6714af17efa6c63) to address this.

fiuczy committed:
mdev: MDevSysfsData.load() without path.exists() check

Remove path.exists() from MDevSysfsData.load() and rely on
fs::canonicalize() used by load_parent_from_sysfs() and
load_mdev_type_from_sysfs() to return an error of kind
std::io::ErrorKind::NotFound. Also adding two tests for a uncovered
scenario.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>

fiuczy commented:
> To me, this `MDevSysfsData` object feels a bit awkward. It feels like it would be nicer if the `MDev` object had separate 'active config' and a 'defined config' members rather than having this standalone object that can be queried and then used to set values on an `MDev` object. Did you try something like that?

I do not think that the raciness problems retrieving consistent data from the sysfs is directly related to how the `MDev` object manages the persisted and active configurations. I do not see why that would be necessary but I suggest to address that issue separately.

fiuczy commented:
It has been a while since I last push changes. How do we want to continue with this?

fiuczy commented:
> > For patch 2 ("add tests with broken active mdevs"), I'm a little bit confused. Did you actually run into broken mdevs in sysfs during normal operation? And if so, was mdevctl doing the wrong thing in this case? This seems to only add tests cases and doesn't seem to modify behavior...
> 
> Yes, patch 2 adds only tests which cover the scenario that the mdev is somehow in an incorrect state. There are no changes in mdevctl behavior itself. These changes are introduced in later patches.
> 
> We have seen problems with the mdevctl code which reads data from sysfs when testing with many mdevs. The scenarios causing errors resulted from mdevctl reading sysfs for mdev x and in parallel the mdev x is deleted. We have not found mdevctl creating invalid mdevs.

To give you an idea what we encounter sometimes in our tests:
2024-04-10 13:46:00.959+0000: 8380: error : virMdevctlModify:958 : internal error: Unable to modify mediated device: ap-check: vfio-ap module does not support ap_config for live modificationError: Script '"/usr/lib/mdevctl/scripts.d/callouts/ap-check"' failed with status '255'
2024-04-10 13:46:02.080+0000: 8443: error : nodeDeviceUpdateMediatedDevices:1923 : internal error: failed to query mdevs from mdevctl: Error: Device must have a defined mdev_type
2024-04-10 13:46:02.080+0000: 8443: warning : udevRemoveOneDeviceSysPath:1631 : Update of mediated device failed
2024-04-10 13:46:02.130+0000: 8399: error : nodeDeviceUpdateMediatedDevices:1923 : internal error: failed to query mdevs from mdevctl: Error: Device must have a defined mdev_type

As you can see the problem is not obvious but by fixing the race condition when reading from sysfs this problem is fixed.

fiuczy commented:
Ping??

jonner commented:
sorry for the long delay. Merged now.
